src/mlpack/core/optimizers/cmaes/cmaes_impl.hpp

```diff
@@ -0,0 +1,699 @@
+<< std::endl;
+
+    diagonalCov = 0; // default is 0, but this might change in future
+
+      xstart.set_size(N);
+      if (startP)
+      xstart = start;
+      else
+      {
+        typicalXcase = true;
+        xstart.fill(0.5);
+      }
+
+    rgInitialStds.set_size(N);
+    if (initDev)
+      rgInitialStds = stdDivs;
+    else
+      rgInitialStds.fill(0.3);
+
+    if (lambda < 2)
+      lambda = 4 + (int) (3.0*log((double) N));
+    if (mu <= 0)
+      mu = lambda / 2;
+
+      weights.set_size(mu);
+      for (int i = 0; i < mu; ++i) weights[i] = log(mu + 1.) - log(i + 1.);
+
+
+      // normalize weights vector and set mueff
+      double s1 = arma::accu(weights);
+      double s2 = arma::accu(weights % weights);
+
+      mueff = s1*s1/s2;
+      weights /= s1;
+
+    if (cs > 0)
+      cs *= (mueff + 2.) / (N + mueff + 3.);
+    if (cs <= 0 || cs >= 1)
+      cs = (mueff + 2.) / (N + mueff + 3.);
+
+    if (ccumcov <= 0 || ccumcov > 1)
+      ccumcov = 4. / (N + 4);
+
+    if (mucov < 1)
+      mucov = mueff;
+    double t1 = 2. / ((N + 1.4142)*(N + 1.4142));
+    double t2 = (2.* mueff - 1.) / ((N + 2.)*(N + 2.) + mueff);
+    t2 = (t2 > 1) ? 1 : t2;
+    t2 = (1. / mucov)* t1 + (1. - 1. / mucov)* t2;
+    if (ccov >= 0)
+      ccov *= t2;
+    if (ccov < 0 || ccov > 1)
+      ccov = t2;
+
+    if (diagonalCov < 0)
+      diagonalCov = 2 + 100. * N / sqrt((double) lambda);
+
+    if (stopMaxFunEvals <= 0)
+      stopMaxFunEvals = facmaxeval * 900 * (N + 3)*(N + 3);
+    else
+      stopMaxFunEvals *= facmaxeval;
+
+    if (iters <= 0)
+      stopMaxIter = ceil((double) (stopMaxFunEvals / lambda));
+    else
+        stopMaxIter = iters;
+
+    if (damps < double(0))
+    {
+    damps = double(1); damps = damps *
+    (double(1) + double(2)*std::max(double(0), std::sqrt((mueff -
+    double(1)) / (N + double(1))) - double(1))) * (double)
+    std::max(double(0.3), double(1) - // modify for short runs
+    (double) N / (double(1e-6) + std::min(stopMaxIter, stopMaxFunEvals
+    / lambda))) + cs;
+    }
+
+    if (updateCmode.modulo < 0)
+      updateCmode.modulo = 1. / ccov / (double) N / 10.;
+    updateCmode.modulo *= facupdateCmode;
+    if (updateCmode.maxtime < 0)
+      updateCmode.maxtime = 0.20;
+  }
+
+  template<typename funcType>
+  double CMAES<funcType>::Optimize(arma::mat& arr)
+  {
+    arFunvals.set_size(lambda);
+    init(arFunvals);
+
+    arma::mat x(N, 1);
+    double funs = 0;
+    double numFun = function.NumFunctions();
+
+  while (!testForTermination())
+  {
+    // Generate lambda new search points, sample population
+    samplePopulation();
+
+    // evaluate the new search points using the given evaluate
+    // function by the user
+    for (int i = 0; i < lambda; ++i)
+    {
+      x = population.submat(i, 0, i, N-1);
+
+      funs = 0;
+      for (ssize_t j = 0; j < numFun; j++)
+      funs += function.Evaluate(x, j);
+
+      arFunvals[i] = funs;
```

Comment:
We could use ``arFunvals[i]`` to store the current fitness, instead of using ``funs`` as extra parameter and we could avoid an extra copy, what do you think?